server/athenian/api/auth.py

```diff
@@ -189,13 +191,14 @@ def __enter__(self):
         def verify_security(auth_funcs, required_scopes, function):
             @functools.wraps(function)
             async def wrapper(request: ConnexionRequest):
-                if "Authorization" not in request.headers or self.force_user:
+                has_auth = "Authorization" in request.headers or "X-API-Key" in request.headers
+                if not has_auth or self.force_user:
                     # Otherwise we will never reach self.extract_token and self._set_user
                     request.headers = CIMultiDict(request.headers)
                     request.headers["Authorization"] = "Bearer null"
                 token_info = get_authorization_info(auth_funcs, request, required_scopes)
```

Comment:
I think it would be cleaner to do as follows so that you don't need to check for specific headers since the security schemes are actually defined in the yaml and handled by `connexion`:

```
    def __enter__(self):
        """Monkey-patch connexion.operations.secure.verify_security()."""
        self._verify_security = secure.verify_security

        def get_auth_info(auth_funcs, request, required_scopes):
            fallback_token = {"token": null, "method": "bearer"}  # this will be still handled correctly by `self._extract_bearer_token`
            if self.force_user:  # no need to get real authorization info
                return fallback_token

            try:
                return get_authorization_info(auth_funcs, request, required_scopes)
            except OAuthProblem:  # nor "Authorization" nor "X-API-Key" headers provided, but no need to check them manullay
                return fallback_token

        def verify_security(auth_funcs, required_scopes, function):
            @functools.wraps(function)
            async def wrapper(request: ConnexionRequest):
                token_info = get_auth_info(auth_funcs, request, required_scopes)
                # token_info = {"token": <token>, "method": <method>} at this point, now do the real work
                await self._set_user(request.context, **token_info)
                return await function(request)
            return wrapper

        secure.verify_security = verify_security
        return self
```

Maybe even

Reply:
We cannot do this:
```
 try:
                return get_authorization_info(auth_funcs, request, required_scopes)
            except OAuthProblem:
                return fallback_token
```
because it means: if the user provided a wrong JWT (e.g., expired), do not return 401 Unauthorized but make them @gkwillie. This is certainly not what we want :D

The current logic is: if we don't supply `Authorization` header, we are @gkwillie. Otherwise, check the token. I'll add comment.

Reply:
I added more comments.

Reply:
Ahh indeed, yes sorry I didn't notice that `OAuthProblem` is not raised only [here](https://github.com/zalando/connexion/blob/f55cb1c923a2b51f402513fc6bdd3c9ebe31353a/connexion/decorators/security.py#L132), otherwise it should have worked fine. Otherwise, with this code we'd need to check for the description which is not very good IMO. It would have been better to me if `connexion` raised a different child error of `OAuthProblem` in case of no auth token provided.

Reply:
This code is certainly a hack with monkey-patching, I don't like it too. I didn't see another way to do what we need back in February, unfortunately.